pandas/tests/indexing/multiindex/test_multiindex.py

```diff
@@ -98,3 +99,33 @@ def test_multiindex_with_datatime_level_preserves_freq(self):
         result = df.loc[0].index
         tm.assert_index_equal(result, dti)
         assert result.freq == dti.freq
+
+    @pytest.mark.parametrize(
+        "mi1",
+        [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ],
+    )
+    @pytest.mark.parametrize(
+        "mi2",
+        [
+            MultiIndex.from_product([["a"], range(2)]),
+            MultiIndex.from_product([["a"], np.arange(2.0)]),
+            MultiIndex.from_product([["b"], ["A", "B"]]),
+            MultiIndex.from_product(
+                [["c"], pd.date_range(start="2017", end="2018", periods=2)]
+            ),
+        ],
+    )
+    def test_concat_with_various_multiindex_dtypes(self, mi1, mi2):
+        # GH #23478
+        df1 = DataFrame(np.zeros((1, len(mi1))), columns=mi1)
+        df2 = DataFrame(np.zeros((1, len(mi2))), columns=mi2)
+
+        with tm.assert_produces_warning(False):
```

Comment:
you're trying to check that the pd.concat call _doesnt_ issue a warning?  use None instead of False

Reply:
I have changed the ```assert_produces_warning``` argument to ```None```.